libs/application-generic/src/utils/sanitize-preview-control-values.ts

```diff
@@ -0,0 +1,222 @@
+const EMPTY_TIP_TAP_OBJECT = JSON.stringify({
+  type: 'doc',
+  content: [
+    {
+      type: 'paragraph',
+      attrs: { textAlign: 'left' },
+      content: [{ type: 'text', text: ' ' }],
+    },
+  ],
+});
+const WHITESPACE = ' ';
+
+type Redirect = {
+  url: string;
+  target: '_self' | '_blank' | '_parent' | '_top' | '_unfencedTop';
+};
+
+type Action = {
+  label?: string;
+  redirect?: Redirect;
+};
+
+type LookBackWindow = {
+  amount: number;
+  unit: string;
+};
+
+function sanitizeRedirect(redirect: Redirect) {
+  if (!redirect.url || !redirect.target) {
+    return undefined;
+  }
+
+  return {
+    url: redirect.url || 'https://example.com',
+    target: redirect.target || '_self',
+  };
+}
+
+function sanitizeAction(action: Action) {
+  if (!action?.label) {
+    return undefined;
+  }
+
+  return {
+    label: action.label,
+    redirect: sanitizeRedirect(action.redirect),
+  };
+}
+
+function sanitizeInApp(controlValues: Record<string, unknown>) {
+  if (!controlValues) return controlValues;
+
+  const normalized: Record<string, unknown> = {
+    subject: controlValues.subject || null,
+    body:
+      (controlValues.body as string)?.length === 0
+        ? WHITESPACE
+        : controlValues.body,
+    avatar: controlValues.avatar || null,
+    primaryAction: null,
+    secondaryAction: null,
+    redirect: null,
+    data: controlValues.data || null,
+  };
+
+  if (controlValues.primaryAction) {
+    normalized.primaryAction = sanitizeAction(
+      controlValues.primaryAction as Action,
+    );
+  }
+
+  if (controlValues.secondaryAction) {
+    normalized.secondaryAction = sanitizeAction(
+      controlValues.secondaryAction as Action,
+    );
+  }
+
+  if (controlValues.redirect) {
+    normalized.redirect = sanitizeRedirect(controlValues.redirect as Redirect);
+  }
+
+  if (typeof normalized === 'object' && normalized !== null) {
+    return Object.fromEntries(
+      Object.entries(normalized).filter(([_, value]) => value !== null),
+    );
+  }
+
+  return normalized;
+}
+
+function sanitizeEmail(controlValues: Record<string, unknown>) {
+  if (!controlValues) return controlValues;
+
+  const emailControls: Record<string, unknown> = {};
+
+  /*
+   * if (controlValues.body != null) {
+   *   emailControls.body = controlValues.body || '';
+   * }
+   */
+  emailControls.subject = controlValues.subject || '';
+  emailControls.body = controlValues.body || EMPTY_TIP_TAP_OBJECT;
+  emailControls.data = controlValues.data || null;
```

Comment:
What is `emailControls.data`? We have a `data` field on `inApp` notifications but not on email.